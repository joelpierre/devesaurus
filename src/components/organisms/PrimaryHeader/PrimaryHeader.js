import React, { createRef, PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Link } from 'gatsby';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { connect } from 'react-redux';
import * as styles from './PrimaryHeader.module.scss';
import Brand from '../../atoms/Brand/Brand';
import Heading from '../../core/Heading/Heading';
import InputField from '../../atoms/InputField/InputField';
import Button from '../../molecules/Buttons/Button';
import Hamburger from '../../atoms/Hamburger/Hamburger';
import * as coreActions from '../../../store/actions/index';

export class PrimaryHeader extends PureComponent {
  constructor() {
    super();

    this.state = {
      searchFormData: '',
    };

    this.searchForm = createRef();

    this.toggleMenu = this.toggleMenu.bind(this);
    this.handleSearchInputOnChange = this.handleSearchInputOnChange.bind(this);
    this.onSearchFormSubmitHandler = this.onSearchFormSubmitHandler.bind(this);
  }

  onSearchFormSubmitHandler(e) {
    e.preventDefault();
    const { searchFormData } = this.state;
    console.log('Form Data: ', searchFormData);
    console.log('Do something on submit');
    this.searchForm.current.reset();
  }

  handleSearchInputOnChange(e) {
    e.preventDefault();
    const {
      target: { value },
    } = e;

    this.setState(() => ({
      searchFormData: value,
    }));
  }

  toggleMenu(e) {
    e.preventDefault();
    const { isMenuOpen, setMenuState } = this.props;
    setMenuState(!isMenuOpen);
  }

  render() {
    const { button: buttonData, input: inputData } = searchFormData;
    const { title, isMenuOpen, alt } = this.props;

    return (
      <header
        data-test="component-primary-header"
        className={classNames(styles['primary-header'], {
          [styles['primary-header--alt']]: alt,
        })}
      >
        <div className={styles['primary-header__wrapper']}>
          <Hamburger
            classes={styles['primary-header__hamburger']}
            descriptor="Menu"
            alt
            isMenuOpen={isMenuOpen}
            onClick={this.toggleMenu}
          />

          <Link to="/">
            <Brand
              type="symbol-inv"
              left
              classes={styles['primary-header__brand']}
            />
          </Link>

          {title && (
            <div className={styles['primary-header__title']}>
              <Heading
                priority={1}
                classes={styles['primary-header__title-text']}
              >
                {title}
              </Heading>
            </div>
          )}

          <div className={styles['primary-header__search']}>
            <form
              ref={this.searchForm}
              className={styles['primary-header__search-form']}
              onSubmit={this.onSearchFormSubmitHandler}
            >
              {inputData && (
                <InputField
                  {...inputData}
                  onChange={this.handleSearchInputOnChange}
                />
              )}
              {buttonData && <Button {...buttonData}>{buttonData.text}</Button>}
            </form>
          </div>
        </div>
      </header>
    );
  }
}

const searchFormData = {
  button: {
    behavior: 'action',
    theme: 'brand',
    type: 'submit',
    text: (
      <FontAwesomeIcon
        icon={['fas', 'search']}
        className={styles['primary-header__search-icon']}
      />
    ),
    classes: styles['primary-header__search-btn'],
  },
  input: {
    classes: styles['primary-header__search-input'],
    name: 'hero-search',
    type: 'text',
    placeholder: 'Enter a search term. e.g. HTML',
  },
};

PrimaryHeader.defaultProps = {
  title: undefined,
  alt: false,
};

PrimaryHeader.propTypes = {
  title: PropTypes.string,
  alt: PropTypes.bool,
  isMenuOpen: PropTypes.bool.isRequired,
  setMenuState: PropTypes.func.isRequired,
};

/* istanbul ignore next */
const mapStateToProps = ({ core: { isMenuOpen } }) => ({
  isMenuOpen,
});

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  setMenuState: data => dispatch(coreActions.setMenuState(data)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PrimaryHeader);
