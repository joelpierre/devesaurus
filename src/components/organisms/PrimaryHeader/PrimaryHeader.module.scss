@import "~sass/core";

.primary-header {
  $root: &;

  height: auto;
  align-items: center;
  top: 0;
  left: 0;
  z-index: 999999;
  width: 100%;
  background-color: $color-tint-alpha;
  padding: 1rem 0;
  transition: $transition-all-base;

  &__wrapper {
    display: flex;
    flex-wrap: nowrap;
    flex-basis: 100%;
    min-width: 0;
    align-content: center;
  }

  &__brand {
    width: auto;
    height: auto;
    flex: 1;
    text-align: left;

    @include tablet-landscape {
    }
  }

  &__menu {
    display: none;

    @include tablet-landscape {
      display: flex;
      flex: 1;
      margin: 0 auto;
    }
  }

  &__hamburger {
    display: flex;

    @include tablet-landscape {
      display: none;
    }
  }
}
