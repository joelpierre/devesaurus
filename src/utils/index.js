export const updateObject = (oldObject, updatedValues) => ({
  ...oldObject,
  ...updatedValues,
});

export const mapOverACFComponents = components => {
  components.map(component => {
    component.acf_fc_layout = component.acf_fc_layout.replace(
      'WordPressAcf_',
      ''
    );
  });
  return components;
};

export const sortWordObj = word => {
  if (word) {
    delete word.author_id;
    delete word.author;
    delete word.author_nicename;
    delete word.content;
    delete word.date_modified;
    delete word.excerpt;
    delete word.yoast;

    return word;
  }

  return false;
};

export const randTheme = () => {
  const theme = ['brand', 'alpha', 'beta', 'gamma', 'psi', 'omega'];
  const rand = randNumBetween(0, theme.length);
  return theme[rand];
};

export const randNumBetween = (min = 0, max = 10) => {
  return Math.floor(Math.random() * max) + min;
};

export const mapTaxonomyTheme = (slug = 'default') => {
  let theme;

  switch (slug) {
    case 'development':
    case 'web-development':
      theme = 'alpha';
      break;
    case 'graphic-design':
    case 'design':
      theme = 'beta';
      break;
    case 'css':
    case 'css3':
      theme = 'gamma';
      break;
    case 'html':
    case 'html5':
      theme = 'psi';
      break;
    case 'news':
      theme = 'omega';
      break;
    default:
      theme = 'brand';
  }

  return theme;
};

export const breakpoint = {
  is(s) {
    const size = s.trim();
    const sizes = {
      xsmall: '599px',
      small: '600px',
      medium: '900px',
      large: '1200px',
      xlarge: '1800px',
    };

    // console.log(window.matchMedia);

    // Eslint doesn't like you accessing hasOwnProperty directly on object.
    // https://eslint.org/docs/rules/no-prototype-builtins
    if (Object.prototype.hasOwnProperty.call(sizes, size)) {
      return window.matchMedia(`only screen and (min-width: ${sizes[size]})`)
        .matches;
    }

    throw new ReferenceError(
      `The size ${size} is not a valid breakpoint: ${JSON.stringify(sizes)}`
    );
  },
};
