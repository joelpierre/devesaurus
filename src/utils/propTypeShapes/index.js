import PropTypes from 'prop-types';

export const wordPropTypeShape = {
  word: PropTypes.string.isRequired,
  origin: PropTypes.string.isRequired,
  syllables: PropTypes.shape({
    count: PropTypes.number.isRequired,
    list: PropTypes.instanceOf(Array).isRequired,
  }).isRequired,
  pronunciation: PropTypes.string,
  results: PropTypes.arrayOf(PropTypes.shape({
    definition: PropTypes.string.isRequired,
    partOfSpeech: PropTypes.oneOf(['noun', 'verb', 'adjective', 'pronoun', 'adverb', 'preposition', 'conjunction', 'interjection']),
  })).isRequired,
  category: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
  }),
  tags: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
  })),
};
