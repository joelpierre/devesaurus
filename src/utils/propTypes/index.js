import PropTypes from 'prop-types';

export const wordPropTypeShape = {
  title: PropTypes.string.isRequired,
  slug: PropTypes.string.isRequired,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      slug: PropTypes.string.isRequired,
      taxonomy: PropTypes.string.isRequired,
    })
  ),
  acf: PropTypes.shape({
    definition: PropTypes.string,
    origin: PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    }),
    syllables: PropTypes.shape({
      count: PropTypes.string.isRequired,
      list: PropTypes.arrayOf(
        PropTypes.shape({
          item: PropTypes.string,
        })
      ).isRequired,
    }).isRequired,
    pronunciation: PropTypes.string,
  }),
};

export const columnShape = PropTypes.oneOf([
  '1',
  1,
  '2',
  2,
  '3',
  3,
  '4',
  4,
  '5',
  5,
  '6',
  6,
  '7',
  7,
  '8',
  8,
  '9',
  9,
  '10',
  10,
  '11',
  11,
  '12',
  12,
]);

export const defaultSize = (size = 'md') => ({
  size,
});

export const sizePropType = {
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', 'full']),
};

export const defaultTheme = (theme = 'brand') => ({
  theme,
});

export const themePropType = {
  theme: PropTypes.oneOf([
    'brand',
    'alpha',
    'beta',
    'gamma',
    'psi',
    'omega',
    'tint-alpha',
    'tint-beta',
    'tint-gamma',
    'tint-psi',
    'tint-omega',
    'gradient-brand',
  ]),
};

export const defaultPageTheme = (theme = 'brand') => ({
  pageTheme: theme,
});

export const pageThemePropType = {
  pageTheme: PropTypes.oneOf([
    'brand',
    'alpha',
    'beta',
    'gamma',
    'psi',
    'omega',
    'tint-alpha',
    'tint-beta',
    'tint-gamma',
    'tint-psi',
    'tint-omega',
  ]),
};
