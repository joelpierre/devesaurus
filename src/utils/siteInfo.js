import React from 'react';
import { graphql, StaticQuery } from 'gatsby';

const SiteInfo = () => (
  <StaticQuery
    query={graphql`
      {
        allWordpressSiteMetadata {
          edges {
            node {
              name
              description
            }
          }
        }
      }
    `}
    render={({ allWordpressSiteMetadata }) => (
      <>
        {allWordpressSiteMetadata.edges[0].node.name}
        <span>{allWordpressSiteMetadata.edges[0].node.description}</span>
      </>
    )}
  />
);

export default SiteInfo;
