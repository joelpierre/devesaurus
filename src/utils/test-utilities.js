import checkPropTypes from 'check-prop-types';
import { applyMiddleware, createStore } from 'redux';

import rootReducer from '../store/reducers';
// import { sagaMiddleware } from '../store/configureStore';

// /**
//  * Store Factory
//  * @param initialState
//  * @returns {Store<any, Action> & *}
//  */
// export const storeFactory = (initialState) => {
//   const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
//   return createStoreWithMiddleware(rootReducer, initialState);
// };


/**
 * Return node(s) with the given data-test attribute.
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper.
 * @param {string} val - Value of the data-test attribute.
 * @returns {ShallowWrapper}
 */
export const findByTestAttr = (wrapper, val) => {
  return wrapper.find(`[data-test="${val}"]`);
};


/**
 * Check the props for a component
 * @param component
 * @param conformingProps
 */
export const checkProps = (component, conformingProps) => {
  const propError = checkPropTypes(
    component.propTypes,
    conformingProps,
    'prop',
    component.name,
  );
  expect(propError)
    .toBeUndefined();
};
