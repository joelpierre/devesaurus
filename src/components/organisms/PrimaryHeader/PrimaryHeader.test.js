import React from 'react';
import { shallow } from 'enzyme';

import { findByTestAttr } from '../../../utils/test-utilities';

import PrimaryHeader from './PrimaryHeader';

const defaultProps = {};

/**
 * Factory function to create a ShallowWrapper for the PrimaryHeader component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
  const setupProps = { ...defaultProps, ...props };
  return shallow(<PrimaryHeader {...setupProps}/>);
};

describe('<PrimaryHeader/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('Renders the PrimaryHeader Component without errors', () => {
    const component = findByTestAttr(wrapper, 'component-primary-header');
    expect(component.length)
      .toBe(1);
  });


});
