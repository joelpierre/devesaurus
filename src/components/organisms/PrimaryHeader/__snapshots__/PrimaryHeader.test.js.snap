// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrimaryHeader/> should match snapshot 1`] = `
<header
  className=""
  data-test="component-primary-header"
>
  <div>
    <Component
      alt={true}
      descriptor="Menu"
      isMenuOpen={false}
      onClick={[Function]}
    />
    <mockConstructor
      to="/"
    >
      <Brand
        center={false}
        classes={null}
        left={true}
        right={false}
        type="symbol-inv"
      />
    </mockConstructor>
    <div>
      <form
        onSubmit={[Function]}
      >
        <InputField
          classes={null}
          disabled={false}
          label={null}
          lg={false}
          max={null}
          md={false}
          min={null}
          name="hero-search"
          onChange={[Function]}
          placeholder="Enter a search term. e.g. HTML"
          sm={false}
          type="text"
        />
        <Button
          behavior="action"
          caps={true}
          size="md"
          text={
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "far",
                  "search",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              symbol={false}
              title=""
              transform={null}
            />
          }
          theme="brand"
          type="submit"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Array [
                "far",
                "search",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
      </form>
    </div>
  </div>
</header>
`;
